packager/VisibilityChecker.cc

```diff
@@ -0,0 +1,558 @@
+                auto file = packageSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    packageSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+
+        {
+            auto testSym = core::Symbols::root();
+            auto member = testSym.data(ctx)->findMember(ctx, core::Names::Constants::Test());
+            if (!member.exists() || !member.isClassOrModule()) {
+                return;
+            }
+
+            testSym = member.asClassOrModuleRef();
+            for (auto name : names) {
+                auto member = testSym.data(ctx)->findMember(ctx, name);
+                if (!member.exists() || !member.isClassOrModule()) {
+                    testSym = core::Symbols::noClassOrModule();
+                    break;
+                }
+
+                testSym = member.asClassOrModuleRef();
+            }
+
+            if (testSym.exists()) {
+                auto file = testSym.data(ctx)->loc().file();
+                if (db.getPackageNameForFile(file) != this->package.mangledName()) {
+                    testSym.data(ctx)->addLoc(ctx, ctx.locAt(loc));
+                }
+            }
+        }
+    }
+
+    void checkExportPackage(core::MutableContext &ctx, core::LocOffsets loc, core::NameRef litPackage) {
+        if (litPackage != this->package.mangledName() && !this->package.importsPackage(litPackage)) {
+            if (auto e = ctx.beginError(loc, core::errors::Packager::InvalidExport)) {
+                e.setHeader("Exporting a symbol that isn't owned by this package");
+            }
+        }
+    }
+
+    PropagateVisibility(const core::packages::PackageInfo &package) : package{package} {}
+
+public:
+    // Find uses of export and mark the symbols they mention as exported.
+    ast::ExpressionPtr postTransformSend(core::MutableContext ctx, ast::ExpressionPtr tree) {
+        auto &send = ast::cast_tree_nonnull<ast::Send>(tree);
+        if (send.fun != core::Names::export_()) {
+            return tree;
+        }
+
+        if (send.numPosArgs() != 1) {
+            // an error will have been raised in the packager pass
+            return tree;
+        }
+
+        auto *lit = ast::cast_tree<ast::ConstantLit>(send.getPosArg(0));
+        if (lit == nullptr || lit->symbol == core::Symbols::StubModule()) {
+            // We don't raise an explicit error here, as this is one of two cases:
+            //   1. Export is given a non-constant argument
+            //   2. The argument failed to resolve
+            // In both cases, errors will be raised by previous passes.
+            return tree;
+        }
+
+        auto &db = ctx.state.packageDB();
+        if (lit->symbol.isClassOrModule()) {
+            auto sym = lit->symbol.asClassOrModuleRef();
+            checkExportPackage(ctx, send.loc, db.getPackageNameForFile(sym.data(ctx)->loc().file()));
```

Comment:
Were there any follow up changes we wanted to try with `getPackageNameForFile`?

It seems like in this case the API we actually want is to go from a symbol to the package that owns it, and we're doing that in an indirect way right now by way of files and mangled packages names. Not sure if we tried changing this and couldn't, put it off doing it, or was never something we wanted to change from the beginning.

Reply:
I didn't push too hard on that because it felt like adding a `NameRef` to every symbol would be expensive in terms of memory. I also think that would be the best possible api for going from a symbol to the package that owns it, and the current implementation was a bit of a compromise to not add too much overhead for users that aren't Stripe.